### PR TITLE
refactor(ivy): remove dynamicViewCount from LContainer

### PR DESCRIPTION
**Pull request by marclaval**
From my understanding, the `dynamicViewCount` property of `LContainer` was introduced during the initial implementation of `ViewContainerRef` to count the number of dynamic views in the container, but it is now used as a flag to know if the `LContainer` belongs to a `ViewContainerRef` or if it was created from an instruction.
In fact, this flag can be simplified and be simply set when the `LContainer` is created.

Following a tricky suggestion from @pkozlowski-opensource , having a dedicated flag is not even needed as the `nextIndex` property is not used when the `LContainer` belongs to a `ViewContainerRef`. **Using -1 as a special value for `nextIndex` is enough to create a flag**

### CODE REVIEW

File: packages/core/src/render3/interfaces/container.ts [hunk 1]
```diff
@@ -18,6 +18,8 @@ export interface LContainer {
   /**
    * The next active index in the views array to read or write to. This helps us
    * keep track of where we are in the views array.
+   * In the case the LContainer is created for a ViewContainerRef,
+   * it is set to -1 to identify this scenario.
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/container.ts]: Could you add some more reasoning as to why `ViewContainerRef` does not need `nextIndex`? Not obvious from reading this.

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
LGTM, please address one doc comment. Also I would like to hear from @alxhub since he originally implemented this.

File: packages/core/src/render3/interfaces/container.ts (hunk 1 above)
Comment (marclaval) [reply packages/core/src/render3/interfaces/container.ts]: Done

**Review by alxhub — DISMISSED** [review]

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -1439,20 +1438,28 @@ function generateInitialInputs(
...
+ */
 export function createLContainer(
-    parentLNode: LNode, currentView: LView, template?: ComponentTemplate<any>): LContainer {
+    parentLNode: LNode, currentView: LView, template?: ComponentTemplate<any>,
+    isForViewContainerRef?: boolean): LContainer {
   ngDevMode && assertNotNull(parentLNode, 'containers should have a parent');
   return <LContainer>{
     views: [],
-    nextIndex: 0,
+    nextIndex: isForViewContainerRef ? -1 : 0,
```
[the remark this answers is not in the document]
Comment (marclaval) [reply packages/core/src/render3/instructions.ts]: Changed to `null`, this is documented in the LContainer interface

**Comment by marclaval** [comment]
Rebased, PTAL

**Comment by mhevery** [comment]
http://test/OCL:198484568:BASE:198484580:1527640248517:137a13c7

**Review by mhevery — APPROVED** [review]

**Review by mhevery — CHANGES_REQUESTED** [review]
pullapprove test

**Review by mhevery — APPROVED** [review]

**Review by jasonaden — APPROVED** [review]

**Comment by mhevery** [comment]
TO: Caretaker

This PR has been approved but [pullapprove](https://pullapprove.com/angular/angular/pull-request/23963/reviews/#group:build-and-ci) is showing that it has not. No idea why since I am in the group. Could caretaker merge it even thought it is not green.

**Comment by vicb** [comment]
It's only that the bot was taking a nape. Everything is green now.

**Pull request closed without merging**